### PR TITLE
make is_power_of_two a const function

### PR DESCRIPTION
**Pull request by tspiteri**
This makes `is_power_of_two` a const function by using `&` instead of short-circuiting `&&`; Rust supports bitwise `&` for `bool` and short-circuiting is not required in the existing expression.

I don't think this needs a const-hack label as I don't find the changed code less readable, if anything I prefer that it is clearer that short circuiting is not used.

@oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
This change may have performance implications, though maybe even be an improvement since it's now branchless code (though likely llvm figured this out by itself anyway).

@rust-lang/libs are you fine with insta stabilizing the constness here?

**Comment by pitdicker** [comment]
The result in compiler explorer with `&` and `&&` is the same: https://godbolt.org/z/S30HDY

**Comment by tspiteri** [comment]
Yeah, LLVM generates the same IR for both.

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

To confirm the insta-stable const-ness...

**Comment by Centril** [comment]
👋 @Amanieu @Kimundi @withoutboats

**Comment by RalfJung** [comment]
> I don't think this needs a const-hack label as I don't find the changed code less readable, if anything I prefer that it is clearer that short circuiting is not used.

The purpose of the const-hack label is to mark changes that we'll want to revert later, once the original code works in `const fn`.

So, will we want to revert this? My personal thinking is yes; `&` on `bool` is unconventional and (except for const-qualif) there is no reason to use it here. So I am marking this as const-hack; once we do the revert we can still discuss if we wouldn't rather keep the code as it is then.

**Comment by tspiteri** [comment]
Or maybe `self.count_ones() == 1`. (Incidentally, with optimizations LLVM recognizes this pattern in all of the above implementations, so that the IR is a call to `@llvm.ctpop` followed by a comparison to 1.)

**Comment by tspiteri** [comment]
Once FCP for the insta-stable const-ness is over, I could update the method to simply return `self.count_ones() == 1` if required.

(LLVM produces identical IR so there should be no performance issue, and I think it is more readable than the current and the modified versions, which are basically doing the same thing using some bitwise hackery. And `count_ones` is already const, so this would result in a const function without the need for const-hack.)

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -3749,8 +3749,8 @@ assert!(!10", stringify!($SelfT), ".is_power_of_two());", $EndFeature, "
 ```"),
             #[stable(feature = "rust1", since = "1.0.0")]
             #[inline]
-            pub fn is_power_of_two(self) -> bool {
-                (self.wrapping_sub(1)) & self == 0 && !(self == 0)
+            pub const fn is_power_of_two(self) -> bool {
+                ((self.wrapping_sub(1)) & self == 0) & !(self == 0)
```
Comment (lzutao) [inline src/libcore/num/mod.rs]:
```suggestion
                (self.wrapping_sub(1) & self == 0) & (self != 0)
```

**Comment by tspiteri** [comment]
@oli-obk Now that the RFC for the insta-stable constness is over, should I change the body of the function to `self.count_ones() == 1` (which I find most readable and would, I think, make const-hack unnecessary), or accept the style nits fix suggested by @lzutao?

**Comment by oli-obk** [comment]
I like the `count_ones` solution the best.

**Pull request merged**